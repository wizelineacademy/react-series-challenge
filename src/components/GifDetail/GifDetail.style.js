import styled from 'styled-components'

export const GifWrapper = styled.div`
  width: 100%;
  height: 100%;
`
export const GifDetailWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: fit-content;
  height: fit-content;
  margin: auto;
`
export const StyledP = styled.p`
  color: white;
  text-align: center;
`
export const BackButton = styled.span`
  color: white;
  font-size: 24px;
  margin-left: 24%;
  cursor: pointer;

  :hover {
    opacity: 0.85;
  }
`
BackButton.displayName = 'BackButton';
