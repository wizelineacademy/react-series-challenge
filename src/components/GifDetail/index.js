import React, { Component } from 'react'
import {
  GifWrapper,
  GifDetailWrapper,
  StyledP,
  BackButton
} from './GifDetail.style'
import { withRouter } from "react-router-dom";
import FavoriteButton from '../FavoriteButton'

class GifDetail extends Component {
  componentWillMount() {
    this.props.searchGif(this.props.match.params.id)
  }

  renderGifIfExist() {
    const { gif, isFavorite, setFavorite} = this.props
    if(this.props.gif.title) {
      return(
        <GifDetailWrapper>
          <FavoriteButton gif={gif} isFavorite={isFavorite} setFavorite={setFavorite} />
          <img alt={gif.title} src={gif.images.downsized_large.url} />
          <StyledP>{gif.title}</StyledP>
        </GifDetailWrapper>
      )
    } else {
      return (
        <div>
          This gif does not exist
        </div>
      )
    }
  }

  render() {
    return(
      <GifWrapper>
        <BackButton onClick={() => this.props.history.goBack()}>Back</BackButton>
        {this.renderGifIfExist()}
      </GifWrapper>
    )
  }
}

export default withRouter(GifDetail)
