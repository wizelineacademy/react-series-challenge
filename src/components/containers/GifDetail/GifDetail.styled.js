import styled from 'styled-components';

const GifDetailContainer = styled.div`
  margin: 0px 0px 0px 0px;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-flex-flow: row wrap;
  justify-content: flex-end;

  @media (max-width: 500px) {
    flex-direction: column;
  }

  @media (max-width: 800px) {
    justify-content: space-around;
  }
`;

export {
  GifDetailContainer
};
