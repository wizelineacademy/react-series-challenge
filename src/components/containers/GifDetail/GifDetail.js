import React, { Component } from 'react';
import { GifDetailContainer } from './GifDetail.styled';
import {
  GiphyListItemHeader,
  GiphyListItemBody,
  GiphyListItemFooter,
  GiphyLinkFavorite,
  GiphyHeartFavorite} from '../../presentational/GiphyList';
import { connect } from 'react-redux';
import trendingGifsActions from '../../../actions/trendingGifs';
import favoriteGifsActions from '../../../actions/favoriteGifs';
const { creators } = trendingGifsActions;
const { getGifRequested } = creators;
const { creators: creatorsFav } = favoriteGifsActions;
const { addFavoriteGifRequested, removeFavoriteGifRequested } = creatorsFav;

class GifDetail extends Component {
  componentDidMount () {
    const { getGifRequested } = this.props;
    getGifRequested(this.props.gifID);
  }

  render() {
    const {selectedGif: source, addFavoriteGifRequested, removeFavoriteGifRequested} = this.props;

    if (!source.fetching) {
      /*eslint no-script-url:0*/
      return (
        <GifDetailContainer>
          <GiphyListItemHeader>
            <h3>{source.title}</h3>
          </GiphyListItemHeader>
          <GiphyListItemBody>
            <img src={source.images.original.url} alt={source.title} />
          </GiphyListItemBody>
          <GiphyListItemFooter>
            <GiphyLinkFavorite href='javascript:void(0)' onClick={() => {
              if (source.isFavorite) {
                removeFavoriteGifRequested(source.id);
              } else {
                addFavoriteGifRequested(source);
              }
            }}>
            {source.isFavorite === true
              ? <span>Remove from my favorites</span> 
              : <span>Send to my favorites</span>}
            <GiphyHeartFavorite color={source.isFavorite ? 'tomato' : 'lightgray'} />
            </GiphyLinkFavorite>
          </GiphyListItemFooter>
        </GifDetailContainer>
      );
    } else {
      return (
        <GifDetailContainer>
          Loading ...
        </GifDetailContainer>
      );
    }
  }
}

const mapStateToProps = (state) => {
  const {selectedGif} = state.trendingGifs;

  return {
    selectedGif
  };
};

export default 
  connect(
    mapStateToProps,
    {
      getGifRequested,
      addFavoriteGifRequested,
      removeFavoriteGifRequested
    }
  )(GifDetail);
