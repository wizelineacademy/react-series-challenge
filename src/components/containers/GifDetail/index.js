import GifDetail from './GifDetail';

export default GifDetail;