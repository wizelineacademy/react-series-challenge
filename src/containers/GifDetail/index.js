import { connect } from 'react-redux'
import favoritesActions from '../../actions/favorites'
import searchActions from '../../actions/search'
import GifDetail from '../../components/GifDetail'

const mapStateToProps = (state) => ({
  gif: state.activeGif,
  isFavorite: (state.favoriteGifs.filter(gif => gif.id === state.activeGif.id).length > 0) ? true : false
})

const mapDispatchToProps = dispatch => ({
  setFavorite: (gif) => dispatch(favoritesActions.creators.setFavorite(gif)),
  searchGif: (gif) => dispatch(searchActions.creators.searchGif(gif))

})

const GifDetailContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(GifDetail)

export default GifDetailContainer
