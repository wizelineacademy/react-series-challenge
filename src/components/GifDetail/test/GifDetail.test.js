import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import GifDetail from '..';
const props = {
  isFavorite: true,
  gif: {
    title:'title',
    images: {
      downsized_large: {
        url: 'test'
      }
    }
  },
  searchGif: jest.fn(),
  setFavorite: jest.fn(),
  history: {
    goBack: jest.fn()
  },
  match: {
    params: {
      id: 1
    }
  }
}

describe('GifDetail', () => {
  it('should render correctly', () => {
    const output = shallow(
      <GifDetail.WrappedComponent {...props} />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('should be defined', () => {
   expect(GifDetail).toBeDefined();
  });
  it('should render div when no gif', () => {
    const wrapper = shallow(<GifDetail.WrappedComponent {...props} gif={{}} />)
    expect(wrapper.find('div').length).toBe(1)
  });
  it('should call back history', () => {
    const wrapper = shallow(<GifDetail.WrappedComponent {...props} />)
    wrapper.find('BackButton').simulate('click')
    expect(props.history.goBack).toHaveBeenCalled();
  });
});
