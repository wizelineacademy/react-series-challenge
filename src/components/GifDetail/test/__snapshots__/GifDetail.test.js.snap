// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifDetail should render correctly 1`] = `
<styled.div>
  <BackButton
    onClick={[Function]}
  >
    Back
  </BackButton>
  <styled.div>
    <FavoriteButton
      gif={
        Object {
          "images": Object {
            "downsized_large": Object {
              "url": "test",
            },
          },
          "title": "title",
        }
      }
      isFavorite={true}
      setFavorite={[Function]}
    />
    <img
      alt="title"
      src="test"
    />
    <styled.p>
      title
    </styled.p>
  </styled.div>
</styled.div>
`;
